### PR TITLE
Add extra color methods

### PR DESCRIPTION
**Pull request by eunomie**
- `saturateColor(color, amount)`
- `desaturateColor(color, amount)`
- `greyscale(color)`
- `lighten(color, amount)`
- `darken(color, amount)`
- `spin(color, hue_angle)`

`*Color` are a bit redundant but `saturate` is a recognized css functions in [`CssFunctionNode`](https://github.com/google/closure-stylesheets/blob/master/src/com/google/common/css/compiler/ast/CssFunctionNode.java#L95).

Those color manipulation methods are compatible with Less expect `spin` because Less convert to RGB and I only use HSL manipulations.

## Test case

_GSS file to be compiled in test_gss_colors.css_

```css
@def BASE_COLOR #5a7bd8;
.div1 {
 background: BASE_COLOR;
}
.div2 {
 background: lighten(BASE_COLOR, 10);
}
.div3 {
 background: darken(BASE_COLOR, 10);
}
.div4 {
 background: saturateColor(BASE_COLOR, 20);
}
.div5 {
 background: desaturateColor(BASE_COLOR, 20);
}
.div6 {
 background: greyscale(BASE_COLOR);
}
.div7 {
 background: spin(BASE_COLOR, 30);
}
.div8 {
 background: spin(BASE_COLOR, -30);
}
```

```html
<html>
       <head>
               <title>Gss test</title>
               <style>
                       div {
                               width: 400px;
                               height: 100px;
                               text-align: center;
                               font-family: arial;
                       }
               </style>
               <link href="test_gss_colors.css" rel="stylesheet" />
       </head>
       <body>
               <div class="div1"><pre>
@def BASE_COLOR #5a7bd8;
{ background: BASE_COLOR; }</pre></div>
               <div class="div2"><pre>
{ background: lighten(BASE_COLOR, 10); }
{ background: #839ce2; }</pre></div>
               <div class="div3"><pre>
{ background: darken(BASE_COLOR, 10); }
{ background: #315ace; }</pre></div>
               <div class="div4"><pre>
{ background: saturateColor(BASE_COLOR, 20); }
{ background: #4671ec; }</pre></div>
               <div class="div5"><pre>
{ background: desaturateColor(BASE_COLOR, 20); }
{ background: #6e85c4; }</pre></div>
               <div class="div6"><pre>
{ background: greyscale(BASE_COLOR); }
{ background: #999999; }</pre></div>
               <div class="div7"><pre>
{ background: spin(BASE_COLOR, 30); }
{ background: #785ad8; }</pre></div>
               <div class="div8"><pre>
{ background: spin(BASE_COLOR, -30); }
{ background: #5abad8; }</pre></div>
       </body>
</html>
```

See attached file for result.
prilvwhrhbg7b5ptya030legyu5x3ofwx2z2aasspgk

### CODE REVIEW

**Comment by iflan** [comment]
Awesome! Thank you. I'm going through your change now.

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 1]
```diff
@@ -370,7 +375,7 @@ public String getCallResultString(List<String> args)
           baseColorString, args.get(1), args.get(2), args.get(3));
     }
 
-    protected String addHsbToCssColor(
+    public static String addHsbToCssColor(
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: Why the change to `public static`? The idea with `protected` is to indicate that it's *not* part of the public interface of the class for callers, but can be used by subclasses. I know that there is the `parseBoolean` method above, but I don't think it should be `public static` either. What do you think?

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 2]
```diff
@@ -404,7 +409,7 @@ protected String addHsbToCssColor(
      * @param brightnessToAdd The amount of brightness to add (can be negative)
      * @return A CSS String representing the new color
      */
-    public String addHsbToCssColor(String baseColorString,
+    public static String addHsbToCssColor(String baseColorString,
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: Same as above.

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 3]
```diff
@@ -433,7 +438,7 @@ public String addHsbToCssColor(String baseColorString,
      * @param brightnessToAdd The amount of brightness to add
      * @return The modified color
      */
-    public Color addValuesToHsbComponents(Color baseColor,
+    public static Color addValuesToHsbComponents(Color baseColor,
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: And again. :-)

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 4]
```diff
@@ -459,6 +464,298 @@ public Color addValuesToHsbComponents(Color baseColor,
   }
 
   /**
+   * Saturate the specified color. First argument is the color, second is the
+   * amount of saturation (from 0 to 100)
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: Is that 0-100% more than the current saturation? Or will 0 essentially make it grey? I think you should be explicit in the docs, if you can be.

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 5]
```diff
@@ -459,6 +464,298 @@ public Color addValuesToHsbComponents(Color baseColor,
...
+          "The first argument must be a CssHexColorNode or a CssLiteralNode.";
+        errorManager.report(
+          new GssError(message, arg1.getSourceCodeLocation()));
+        throw new GssFunctionException(message);
+      }
+      CssNumericNode numeric2;
+      if (arg2 instanceof CssNumericNode) {
+        numeric2 = (CssNumericNode)arg2;
+      } else {
+        String message = "Arguments number 2 must be CssNumericNodes";
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: How about:
```java
String message = "Argument 2 must be a CssNumericNode";
```

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java (hunk 1 above)
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: Ahh, I see that you're re-using this method down on line 503. Hmm. Yes, I would agree that you wouldn't want to have an instance of `AddHsbToCssColor` to color for that. I'm going to have to think a bit on this one.

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 6]
```diff
@@ -459,6 +464,298 @@ public Color addValuesToHsbComponents(Color baseColor,
...
+        String message = "The first argument must be a CssHexColorNode or a CssLiteralNode.";
+        errorManager.report(new GssError(message, arg1
+          .getSourceCodeLocation()));
+        throw new GssFunctionException(message);
+      }
+      CssNumericNode numeric2;
+      if (arg2 instanceof CssNumericNode) {
+        numeric2 = (CssNumericNode) arg2;
+      } else {
+        String message = "Arguments number 2 must be CssNumericNodes";
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: Like above.

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java (hunk 1 above)
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: I was just looking at `BaseBlendColors`. I wonder if you can't do something similar here by making all of the functions that use `AddHsbToCssColor.addHsbToCssColor` inherit from the same base class, including what is today `AddHsbToCssColor`. What do you think?

File: tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java [hunk 7]
```diff
@@ -204,6 +204,96 @@ public void testMakeTranslucent_otherLocale() {
...
+  public void testSaturateColor() throws GssFunctionException {
+    GssFunctions.SaturateColor function = new GssFunctions.SaturateColor();
+    assertEquals("#2F5BD8",
+        function.getCallResultString(ImmutableList.of("#5a7bd8", "20")));
+  }
+
+  public void testSaturateColor_otherLocale() throws GssFunctionException {
+    Locale.setDefault(Locale.FRANCE);
+    try {
+
```
Comment (iflan) [inline tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java]: Extra line.

File: tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java [hunk 8]
```diff
@@ -204,6 +204,96 @@ public void testMakeTranslucent_otherLocale() {
...
+      GssFunctions.SaturateColor function = new GssFunctions.SaturateColor();
+      assertEquals("#2F5BD8",
+        function.getCallResultString(ImmutableList.of("#5a7bd8", "20")));
```
Comment (iflan) [inline tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java]: The `_otherLocale` tests only matter if there is a fractional part to the result. See issue #14.

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java (hunk 1 above)
Comment (eunomie) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: I'm working on it. I'm adding some methods to manipulate colors using HSL color space (which make more easy some operations).
Yes I can extract some functions and inherit from a base class, it'll be better.

**Comment by eunomie** [comment]
I've made some big changes.
I use HSL to manipulate colors, it's really more easier. And this is compatible with Less.
I've extracted all common methods in a base class. I've firstly done the same thing with HSB methods, it's no more necessary but still there. Would you I remove this change?

**Comment by eunomie** [comment]
I've added a `spin` function. Really useful to have analogous or triad colors from a base color. Spin is a rotation of hue, like a rotation in a color wheel.

**Comment by iflan** [comment]
Hi Yves,

I haven't had a chance to look at this today. I'll take another pass at it tomorrow.

Ian

File: src/com/google/common/css/compiler/gssfunctions/ColorUtil.java [hunk 9]
```diff
@@ -28,6 +29,7 @@
   public static final int H = 0;
   public static final int S = 1;
   public static final int B = 2;
+  public static final int L = 2;
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: I presume this is lightness, right?

File: src/com/google/common/css/compiler/gssfunctions/ColorUtil.java [hunk 10]
```diff
@@ -43,6 +45,39 @@ public static Color hsbToColor(float[] inputHsb) {
     return Color.getHSBColor(inputHsb[H], inputHsb[S], inputHsb[B]);
   }
 
+  public static float[] hsbToHsl(float[] inputHsb) {
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: Since the class is package-private (default visibility), it doesn't really make sense to make the methods be public. Can you remove the public from your methods and the existing ones?

And can you add some JavaDoc?

File: src/com/google/common/css/compiler/gssfunctions/ColorUtil.java [hunk 11]
```diff
@@ -43,6 +45,39 @@ public static Color hsbToColor(float[] inputHsb) {
...
+    hsl[L] = lHsl;
+
+    return hsl;
+  }
+
+  public static float[] toHsl(Color color) {
+    return hsbToHsl(toHsb(color));
+  }
+
+  public static Color hslToColor(float[] inputHsl) {
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: This is basically hsbToColor(hslTohsb(inputHsl)), so why not just supply hslTohsb()?

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 12]
```diff
@@ -459,6 +467,454 @@ public Color addValuesToHsbComponents(Color baseColor,
   }
 
   /**
+   * Base implementation of color manipulation functions in HSL color space
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: How about:
```
Abstract base class providing HSL color space manipulation functions.
```

File: src/com/google/common/css/compiler/gssfunctions/GssFunctions.java [hunk 13]
```diff
@@ -459,6 +467,454 @@ public Color addValuesToHsbComponents(Color baseColor,
...
+          Integer.parseInt(hueToAdd),
+          Integer.parseInt(saturationToAdd),
+          Integer.parseInt(lightnessToAdd));
+      } catch (NumberFormatException e) {
+        String message = String.format("Could not parse the integer arguments"
+            + " for the function 'addHslToCssColor'. The list of arguments was:"
+            + " %s, %s, %s, %s. ",
+            baseColorString, hueToAdd, saturationToAdd, lightnessToAdd);
+        throw new GssFunctionException(message);
+      } catch (IllegalArgumentException e) {
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/GssFunctions.java]: If you want to use Java 7's multi-catch to merge the two catch blocks, I'm OK with that.

File: tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java [hunk 14]
```diff
@@ -204,6 +204,54 @@ public void testMakeTranslucent_otherLocale() {
     }
   }
 
+  public void testSaturateColor() throws GssFunctionException {
```
Comment (iflan) [inline tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java]: Can you add a test or two that exercises the wrap-around behavior of hue and the clipping behavior of saturation and lightness?

**Comment by iflan** [comment]
Awesome work! Please let me know when I should look again.

Also, because this is a big change, I'm going to send this through internal review and commit it there, then export it back to this repo. I'll do my utmost to make sure that the change still comes out as authored by you. :-)

File: src/com/google/common/css/compiler/gssfunctions/ColorUtil.java (hunk 9 above)
Comment (eunomie) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: Yes, that's it.
I'll add a comment before each (and to explain why B and L are both '2')

File: src/com/google/common/css/compiler/gssfunctions/ColorUtil.java (hunk 11 above)
Comment (eunomie) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: I prefer read in caller code `hslToColor(hsl)` than `hslToColor(hslToHsb(hsl))` which can be an implementation detail.
But I can add a `hslToHsb` and change `hslToColor` to call `hslToHsb`.
So the code of `ColorUtil` will be more _symmetric_:

- `hsbToHsl` and `hslToHsb`
- `toHsb` and `toHsl`
- `hsbToColor` and `hslToColor`

File: tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java (hunk 14 above)
Comment (eunomie) [inline tests/com/google/common/css/compiler/gssfunctions/GssFunctionsTest.java]: Yeah, sure.

**Comment by eunomie** [comment]
Thanks for all your interesting comments! I'll check tonight or in the weekend and I'll add a comment when everything is ready on my side.

**Comment by eunomie** [comment]
I've added some comments (hope they are comprehensible ;)) and tests.
I've made 2 PR for try/catch and public/package remarks.

File: src/com/google/common/css/compiler/gssfunctions/ColorUtil.java [hunk 15]
```diff
@@ -25,9 +26,22 @@
  */
 class ColorUtil {
 
+  /**
+   * Index of Hue in HSB and HSL array.
```
Comment (iflan) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: BTW, if you want to make this a one-line JavaDoc, that's fine with me:
```java
/** Index of Hue in HSB and HSL array. */
```

Comment (eunomie) [inline src/com/google/common/css/compiler/gssfunctions/ColorUtil.java]: I've updated this part.

**Comment by eunomie** [comment]
The branch is force-pushed against new master (and fix conflicts in `ColorUtil`)

**Comment by iflan** [comment]
Yves, I'm getting this change reviewed internally today.

**Comment by iflan** [comment]
This is merged with 5e681fe82d57942003056786cfb797822ff66502.

**Comment by eunomie** [comment]
Yeah! Thanks a lot!

**Comment by iflan** [comment]
Thank *you*. BTW, I did make a few changes based on our internal review, but they were minor. :-)
